### PR TITLE
chore(typescript): setup Babel, Rollup, ESLint [PART-1]

### PR DESCRIPTION
**Pull request by samouss**
**Summary**

This is the first PR to setup TypeScript into the project. Here is the notable changes:

- TypeScript intall
- allow Babel to transpile TypeScript file
- allow Rollup to resolve / transpile TypeScript file
- allow ESLint to resolve TypeScript file
- move Babel & Rollup dependencies top level

There is at least one thing that we have to setup on the TypeScript side: create the definition file. It can be done later once we have a proper coverage of the project.

### CODE REVIEW

File: packages/react-instantsearch-core/rollup.config.js [hunk 1]
```diff
@@ -17,12 +17,14 @@ const createLicence = name =>
...
   babel({
     exclude: ['../../node_modules/**', 'node_modules/**'],
+    extensions: ['.js', '.ts', '.tsx'],
     rootMode: 'upward',
     runtimeHelpers: true,
   }),
   resolve({
     browser: true,
     preferBuiltins: false,
+    extensions: ['.mjs', '.js', '.ts', '.jsx', '.tsx', '.json'],
```
Comment (Haroenv) [inline packages/react-instantsearch-core/rollup.config.js]: do we actually use `.mjs`?

File: tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    "target": "ESNEXT",
+    "module": "es2015",
+    "moduleResolution": "node",
+    "jsx": "preserve",
+    "noEmit": true,
+    "strict": true,
+    // It's mandatory to disable this strict check at the moment otherwise we
+    // have to type every import we do. The migration is an incremental process
+    // so it happens that we import JavaScript module from TypeScript. The
+    // compiler translates those imports as impliciy any.
```
Comment (Haroenv) [inline tsconfig.json]: add an `@TODO`?

File: packages/react-instantsearch-core/rollup.config.js (hunk 1 above)
Comment (samouss) [reply packages/react-instantsearch-core/rollup.config.js]: Nope but a dependency could use it.

**Comment by samouss** [comment]
> I'd expect a `tsc` call in the test config, are you doing that in another PR?

Yes in another PR because you have without any TypeScript file the built fail.

**Pull request merged**